x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/index.tsx

```diff
@@ -0,0 +1,415 @@
+  );
+
+  const setAppState = useCallback(
+    (newState: { columns: string[]; sort?: string[][] }) => {
+      if (newState.sort) {
+        onSort(newState.sort);
+      } else {
+        const columnsStates = newState.columns.map((columnId) =>
+          getColumnHeader(columnId, defaultUdtHeaders)
+        );
+        dispatch(timelineActions.updateColumns({ id: timelineId, columns: columnsStates }));
+      }
+    },
+    [dispatch, onSort, timelineId]
+  );
+
+  const { onAddColumn, onRemoveColumn, onSetColumns } = useColumns({
+    capabilities,
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    dataView: dataView!,
+    dataViews,
+    setAppState,
+    useNewFieldsApi: true,
+    columns: columnIds,
+    sort: sortingColumns,
+  });
+
+  const onAddFilter = useCallback(
+    (field: DataViewField | string, values: unknown, operation: '+' | '-') => {
+      if (dataView && timelineFilterManager) {
+        const fieldName = typeof field === 'string' ? field : field.name;
+        popularizeField(dataView, fieldName, dataViews, capabilities);
+        const newFilters = generateFilters(
+          timelineFilterManager,
+          field,
+          values,
+          operation,
+          dataView
+        );
+        return timelineFilterManager.addFilters(newFilters);
+      }
+    },
+    [timelineFilterManager, dataView, dataViews, capabilities]
+  );
+
+  const [{ dragging }] = useDragDropContext();
+  const draggingFieldName = dragging?.id;
+
+  const onToggleColumn = useCallback(
+    (columnId: string) => {
+      dispatch(
+        timelineActions.upsertColumn({
+          column: getColumnHeader(columnId, defaultHeaders),
+          id: timelineId,
+          index: 1,
+        })
+      );
+    },
+    [dispatch, timelineId]
+  );
+
+  const isDropAllowed = useMemo(() => {
+    if (!draggingFieldName || columnIds.includes(draggingFieldName)) {
+      return false;
+    }
+    return true;
+  }, [draggingFieldName, columnIds]);
+
+  const onDropFieldToTable = useCallback(() => {
+    if (draggingFieldName) {
+      onAddColumn(draggingFieldName);
+      onToggleColumn(draggingFieldName);
+    }
+  }, [draggingFieldName, onAddColumn, onToggleColumn]);
+
+  const onAddFieldToWorkspace = useCallback(
+    (field: DataViewField) => {
+      onAddColumn(field.name);
+      onToggleColumn(field.name);
+    },
+    [onAddColumn, onToggleColumn]
+  );
+
+  const onRemoveFieldFromWorkspace = useCallback(
+    (field: DataViewField) => {
+      if (columns.some(({ id }) => id === field.name)) {
+        dispatch(
+          timelineActions.removeColumn({
+            columnId: field.name,
+            id: timelineId,
+          })
+        );
+      }
+      onRemoveColumn(field.name);
+    },
+    [columns, dispatch, onRemoveColumn, timelineId]
+  );
+
+  const onFieldEdited = useCallback(() => {
+    refetch();
+  }, [refetch]);
+
+  const wrappedOnFieldEdited = useCallback(async () => {
```

Comment:
why is this async?

Reply:
it is because the type of prop demands an async function.